File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.cpp
class Solution {
public:
    int helper(int idx,string &s,map<string,int> &mp,vector<int> &dp){
        if(idx==s.size()) return 0;
        if(dp[idx]!=-1) return dp[idx]; 
        int ans=100;
        for(int i=idx+1;i<=s.size();i++){
            string tmp=s.substr(idx,i-idx);
            if(mp.find(tmp)!=mp.end()) ans=min(ans,helper(i,s,mp,dp));
            else ans=min(ans,(i-idx)+helper(i,s,mp,dp));
        }
        return dp[idx]=ans;
    }
    int minExtraChar(string s, vector<string>& dictionary) {
        map<string,int> mp;
        for(auto it:dictionary) mp[it]++;
        vector<int> dp(s.size()+1,-1);
        return helper(0,s,mp,dp);
    }
};

File: 2707-extra-characters-in-a-string/2707-extra-characters-in-a-string.java
class Solution {
    HashMap<String,Integer> map;
    int[] dp;
    int helper(int idx,String s){
        if(idx==s.length()) return 0;
        int ans=100;
        if(dp[idx]!=-1) return dp[idx];
        for(int i=idx+1;i<=s.length();i++){
            String tmp=s.substring(idx,i);
            if(map.containsKey(tmp)) ans=Math.min(ans,helper(i,s));
            else ans=Math.min(ans,(i-idx)+helper(i,s));
        }
        return dp[idx]=ans;
    }
    public int minExtraChar(String s, String[] dictionary) {
        map = new HashMap<>();
        dp = new int[100];
        for(int i=0;i<100;i++) dp[i]=-1;
        for(String it:dictionary){
            map.put(it,1);
        }
        
        return helper(0,s);
    }
}